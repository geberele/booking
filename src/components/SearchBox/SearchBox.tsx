import React, { useState } from 'react'
import { getSearchResults, IResponse, IResults } from '../../api/search'
import { Autocomplete } from '../Autocomplete/Autocomplete'

export const SearchBox = () => {

  const [results, setResults] = useState<IResults[]>([])
  const [inputValue, setInputValue] = useState<string>('')

  const handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value
    setInputValue(value)
    if  (value.length >= 2) {
      getSearchResults(e.target.value)
        .then((res: IResponse) => {
          if (!res.hasError) {
            setResults(res.data)
          }
        })
    } else {
      setResults([])
    }
  }

  const handleOnClick = (value: string) => {
    setInputValue(value)
    setResults([])
  }

  return (
    <div className="search-box">
      <h2>Where are you going?</h2>
      <label
        id="search-box-label"
        htmlFor="search-box-input"
      >
        Pick-up Location
      </label>
      <input
        id="search-box-input"
        data-testid="search-box-input"
        className="search-box__input"
        type="text"
        placeholder="city, airport, station, region and district..."
        aria-labelledby="search-box-label"
        onChange={handleOnChange}
        value={inputValue}
      />
      <Autocomplete
        results={results}
        onClick={handleOnClick}
      />
    </div>
  )
}
