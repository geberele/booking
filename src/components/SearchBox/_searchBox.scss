.search-box {
  background: $color-yellow;
  margin: 100px auto;
  padding: 24px;

  @media (min-width: $breakpoint-tablet) {
    width: 512px;
  }
}
