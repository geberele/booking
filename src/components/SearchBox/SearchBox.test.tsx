import * as React from 'react'
import { render, fireEvent, waitForElement } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { SearchBox } from './SearchBox'

describe('SearchBox', () => {
  it('renders', () => {
    const { asFragment } = render(<SearchBox />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('TEST 1 - AC1: shows the correct text for the label', () => {
    const { getByLabelText } = render(<SearchBox />)
    const label = getByLabelText(/Pick-up Location/i)
    expect(label).toBeInTheDocument()
  })

  it('TEST 1 - AC2: shows the correct text for the placeholder', () => {
    const { getByPlaceholderText } = render(<SearchBox />)
    const input = getByPlaceholderText(/city, airport, station, region and district.../i)
    expect(input).toBeInTheDocument()
  })

  it('TEST 1 - AC3: applies the focus state', () => {
    const { getByPlaceholderText } = render(<SearchBox />)
    const input = getByPlaceholderText(/city, airport, station, region and district.../i)
    input.focus()
    expect(document.activeElement).toEqual(input)
  })

  it('TEST 1 - AC4: has the input field accessible from screen readers', () => {
    const { container } = render(<SearchBox />)
    const label = container.querySelector(`[id="search-box-label"]`)
    const input = container.querySelector(`[aria-labelledby="search-box-label"]`)
    expect(label).toHaveTextContent('Pick-up Location')
    expect(input).toBeInTheDocument()
  })

  it('TEST 2 - AC1: does not show placeholder and autocomplete when only one character is typed', () => {
    const { container, getByPlaceholderText } = render(<SearchBox />)
    const input = getByPlaceholderText(/city, airport, station, region and district.../i)
    fireEvent.input(input, { target: { value: 'm' } })

    const autocomplete = container.querySelector(`[class="autocomplete"]`)

    expect(input).toHaveValue('m')
    expect(autocomplete).toBeNull()
  })

  it('TEST 2 - AC2: shows autocomplete when 2 alphanumeric characters are typed', async () => {
    const { getByTestId } = render(<SearchBox />)
    const input = getByTestId('search-box-input')
    fireEvent.input(input, { target: { value: 'mi' } })

    const autocomplete = await waitForElement(() => getByTestId('autocomplete'))

    expect(input).toHaveValue('mi')
    expect(autocomplete).toBeVisible()
  })

  it('TEST 2 - AC3: shows autocomplete with maximum 6 items', async () => {
    const { getByTestId, container } = render(<SearchBox />)
    const input = getByTestId('search-box-input')
    fireEvent.input(input, { target: { value: 'mi' } })

    const autocomplete = await waitForElement(() => getByTestId('autocomplete'))
    const items = container.querySelector('ol')

    expect(items?.childElementCount).toBe(6)
  })

  it('TEST 2 - AC4: shows message "No results found" when no items are found', async () => {
    const { getByText, getByTestId } = render(<SearchBox />)
    const input = getByTestId('search-box-input')
    fireEvent.input(input, { target: { value: 'XXXXXXXX' } })

    const noResultsText = await waitForElement(() => getByText(/No results found/))

    expect(noResultsText).toBeInTheDocument()
  })

  it('TEST 2 - AC5: does not show autocomplete when the characters are removed', async () => {
    const { getByTestId, queryByTestId } = render(<SearchBox />)
    const input = getByTestId('search-box-input')
    fireEvent.input(input, { target: { value: 'mi' } })

    const autocomplete = await waitForElement(() => queryByTestId(/autocomplete/i))
    fireEvent.input(input, { target: { value: '' } })

    expect(queryByTestId(/autocomplete/i)).toBeNull()
  })
})
