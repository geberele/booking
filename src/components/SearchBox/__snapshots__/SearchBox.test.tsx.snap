// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBox renders 1`] = `
<DocumentFragment>
  <div
    class="search-box"
  >
    <h2>
      Where are you going?
    </h2>
    <label
      for="search-box-input"
      id="search-box-label"
    >
      Pick-up Location
    </label>
    <input
      aria-labelledby="search-box-label"
      class="search-box__input"
      data-testid="search-box-input"
      id="search-box-input"
      placeholder="city, airport, station, region and district..."
      type="text"
      value=""
    />
  </div>
</DocumentFragment>
`;
